### ISSUE TITLE
There is no support for non native Golang keys, non standard algorithms and alternate entropy sources.

### ISSUE THREAD

**Issue by dmercer-google** [square/go-jose#193]
tl;dr but I am trying to make a case for why this is important.

There are many instances where the keys, algorithms and the entropy source required to sign a JWT are not supported by Golang and go-jose.  [Issue #25](https://github.com/square/go-jose/issues/25) is an example of one such case.

Some examples of unsupported key types include:
* PKCS11 keys (e.g. those provided by [miekg/pkcs11](https://github.com/miekg/pkcs11)). 
* Keys associated with Hardware Security Modules (HSM) exposing APIs other than PKCS11 (e.g. nCipher ncore keys).  
* ECC keys using [NIST P-192](https://csrc.nist.gov/csrc/media/publications/fips/186/3/archive/2009-06-25/documents/fips_186-3.pdf) (see Appendix D para D2.1)  which is unsupported by Golang.
* Key handles associated with various cloud providers (e.g. AWS, Azure and GCP)

There are instances where, so as to comply with internal security policy, signing must be performed with keys that are supported by Golang and/or go-jose but which employ the use of entropy sources other than ```crypto/rand.Reader```. An example of this is where entropy must be generated through specific interaction with a hardware based entropy source (e.g. an HSM)

go-jose only supports a subset of the signing algorithms called out in the [IANA registry of JSON Web Signature and Encryption Algorithms](https://www.iana.org/assignments/jose/jose.xhtml#web-signature-encryption-algorithms).  Unsupported algorithms of note are ```EdDSA``` and ```none```.

There are instances where algorithms not mentioned in the [IANA registry of JSON Web Signature and Encryption Algorithms](https://www.iana.org/assignments/jose/jose.xhtml#web-signature-encryption-algorithms) are required. 

Some examples of such algorithms are:
* [GOST R 34.10-2001](https://tools.ietf.org/html/rfc5832), 
* [KCDSA](https://en.wikipedia.org/wiki/KCDSA)

[RFC 7515 para 4.1.1](https://tools.ietf.org/html/rfc7515#page-10) explicitly allows for the use of unregistered algorithms providing they do not lead to name collisions with IANA registered algorithms.

Lastly: There exist cases where an organization may want to “roll it's own” crypto.  While generally ill advised, go-jose should not prohibit this.  “Rolling one’s own” is often done to support research or to leverage emerging algorithms (e.g. [XMSS](https://datatracker.ietf.org/doc/rfc8391/), [LMS](https://tools.ietf.org/html/draft-mcgrew-hash-sigs-10), etc.).

go-jose should provide a means to support signing that meets all of the aforementioned cases.

Note:  A pull request that addresses this issue will be created shortly.

**Comment by alokmenghrajani** [comment]
I agree with the general idea of being more flexible when it comes to algorithms/HSMs.

> Keys associated with Hardware Security Modules (HSM) exposing APIs other than PKCS11 (e.g. nCipher ncore keys).

nCiphers do expose a pkcs11 API. Moving a key from one application to another is possible (e.g. from java key to pkcs11 or vice versa), albeit painful.

**Comment by dmercer-google** [comment]
@alokmenghrajani 

The nCipher PKCS11 implementation is an incomplete implementation of v2.2 and is this lacking in several areas, despite the underlying hardware having support for the v2.4 mechanisms and algorithms.  I've a lot of experience with nCipher and have basically given up on their P11 drivers and I now live in ncore hell.

**Comment by csstaub** [comment]
Per discussion in #194, this is already supported on v2 through the OpaqueSigner interface. Hence closing this issue.
